benchexec/resources.py

```diff
@@ -209,145 +170,191 @@ def _get_cpu_cores_per_run0(
                 f"of core {core} are not usable. "
                 f"Please always make all virtual cores of a physical core available."
             )
+    # check if all units of the same hierarchy level have the same number of cores
+    for index in range(len(hierarchy_levels)): # [dict, dict, dict, ...]
+        cores_per_unit = len(next(iter(hierarchy_levels[index].values())))
+        print("cores_per_unit of hierarchy_level ", index," = ", cores_per_unit)
+        if any(len(cores) != cores_per_unit for cores in hierarchy_levels[index].values()):
+            sys.exit(
+                "Asymmetric machine architecture not supported: "
+                "CPUs/memory regions with different number of cores."
+            )
 
-    # Second, compute some values we will need.
-    unit_count = len(cores_of_unit)
-    units = sorted(cores_of_unit.keys())
+    # ( compute some values we will need.)
+
+    # coreLimit_rounded_up (int): recalculate # cores for each run accounting for HT
+    core_size =  len(next(iter(siblings_of_core.values()))) # Wert aus hierarchy_level statt siblings_of_core?
     coreLimit_rounded_up = int(math.ceil(coreLimit / core_size) * core_size)
     assert coreLimit <= coreLimit_rounded_up < (coreLimit + core_size)
 
-    units_per_run = int(math.ceil(coreLimit_rounded_up / unit_size))
-    if units_per_run > 1 and units_per_run * num_of_threads > unit_count:
+
+    # Choose hierarchy level for core assignment
+    chosen_level = 1
+    # move up in hierarchy as long as the number of cores at the current level is smaller than the coreLimit
+    # if the number of cores at the current level is as big as the coreLimit: exit loop 
+    while len (next(iter(hierarchy_levels[chosen_level].values()))) < coreLimit_rounded_up  and chosen_level < len (hierarchy_levels):
+        chosen_level = chosen_level+1
+    print("chosen_level = ",chosen_level)
```

Comment:
Please do not use `print()`, this messes up the output of BenchExec.

Debug output can be made with `logging.debug()`, but it needs to be written in a way that is useful for (experienced) users and other developers of BenchExec to understand.